pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -352,29 +367,36 @@ func (m *mergeQuerier) Select(ctx context.Context, sortSeries bool, hints *stora
 	matchedIDs, filteredMatchers := FilterValuesByMatchers(m.idLabelName, ids, matchers...)
 
 	jobs := make([]string, 0, len(matchedIDs))
-	seriesSets := make([]storage.SeriesSet, len(matchedIDs))
 	for id := range matchedIDs {
 		jobs = append(jobs, id)
 	}
 
+	return m.innerSelectFunc(ctx, jobs, start, m.upstreamQueryWaitDuration, m.upstream, m.idLabelName, m.maxConcurrency, sortSeries, hints, filteredMatchers...)
+}
+
+type InnerSelectFunc func(ctx context.Context, jobs []string, start time.Time, upstreamQueryWaitDuration prometheus.Histogram, upstream MergeQuerierUpstream, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, filteredMatchers ...*labels.Matcher) storage.SeriesSet
+
+func defaultInnerSelectFunc(ctx context.Context, jobs []string, start time.Time, upstreamQueryWaitDuration prometheus.Histogram, upstream MergeQuerierUpstream, idLabelName string, maxConcurrency int, sortSeries bool, hints *storage.SelectHints, filteredMatchers ...*labels.Matcher) storage.SeriesSet {
+	seriesSets := make([]storage.SeriesSet, len(jobs))
+
 	// We don't use the context passed to this function, since the context has to live longer
 	// than the call to ForEachJob (i.e. as long as seriesSets)
 	run := func(_ context.Context, idx int) error {
-		m.upstreamQueryWaitDuration.Observe(time.Since(start).Seconds())
+		upstreamQueryWaitDuration.Observe(time.Since(start).Seconds())
```

Comment:
we can get rid of `upstreamQueryWaitDuration` and `start` by wrapping the `upstream` with our own implementation which tracks these. That way other implementations don't need to remember to record the metric.